stac_fastapi/pgstac/stac_fastapi/pgstac/models/links.py

```diff
@@ -0,0 +1,292 @@
+
+import logging
+from typing import Dict, List, Union
+from urllib.parse import ParseResult, parse_qs, unquote, urlencode, urljoin, urlparse
+
+import attr
+from stac_pydantic.api.extensions.paging import PaginationLink
+from stac_pydantic.shared import Link, MimeTypes, Relations
+from starlette.requests import Request
+
+from stac_fastapi.extensions.third_party.tiles import OGCTileLink
+
+logger = logging.getLogger("uvicorn")
+logger.setLevel(logging.INFO)
+
+# These can be inferred from the item/collection so they aren't included in the database
+# Instead they are dynamically generated when querying the database using the classes defined below
+INFERRED_LINK_RELS = ["self", "item", "parent", "collection", "root"]
+
+
+def filter_links(links: List[Dict]) -> List[Dict]:
+    """Remove inferred links."""
+    return [link for link in links if link["rel"] not in INFERRED_LINK_RELS]
+
+
+def merge_params(url: str, newparams: Dict) -> str:
+    """Merge url parameters."""
+    u = urlparse(url)
+    params = parse_qs(u.query)
+    params.update(newparams)
+    param_string = unquote(urlencode(params, True))
+
+    href = ParseResult(
+        scheme=u.scheme,
+        netloc=u.netloc,
+        path=u.path,
+        params=u.params,
+        query=param_string,
+        fragment=u.fragment,
+    ).geturl()
+    return href
+
+
+@attr.s
+class BaseLinks:
+    """Create inferred links common to collections and items."""
+
+    request: Request = attr.ib()
+
+    @property
+    def base_url(self):
+        """Get the base url."""
+        return str(self.request.base_url)
+
+    @property
+    def url(self):
+        """Get the current request url."""
+        return str(self.request.url)
+
+    def resolve(self, url):
+        """Resolve url to the current request url."""
+        return urljoin(str(self.base_url), str(url))
+
+    def link_self(self) -> Link:
+        """Return the self link."""
+        return Link(rel=Relations.self, type=MimeTypes.json, href=self.url)
+
+    def link_root(self) -> Link:
+        """Return the catalog root."""
+        return Link(rel=Relations.root, type=MimeTypes.json, href=self.base_url)
+
+    def create_links(self) -> List[Union[PaginationLink, Link]]:
+        """Return all inferred links."""
+        links = []
+        for name in dir(self):
+            if name.startswith("link_") and callable(getattr(self, name)):
+                link = getattr(self, name)()
+                if link is not None:
+                    links.append(link)
+        return links
+
+    async def get_links(
+        self, extra_links: List[Union[PaginationLink, Link]] = []
+    ) -> List[Union[PaginationLink, Link]]:
+        """
+        Generate all the links.
+
+        Get the links object for a stac resource by iterating through
+        available methods on this class that start with link_.
+        """
+        if self.request.method == "POST":
+            self.request.postbody = await self.request.json()
```

Comment:
```suggestion
            self.request.postbody = await self.request.body()
```

This is a code smell but a problem for another day 😆